### PR TITLE
KEYCLOAK-19570 Add annotation processing for HotRod clients

### PR DESCRIPTION
**Pull request by mhajas**
This PR adds the possibility to automatically generate Map interfaces for HotRod entities automatically.

Here is an example of generated ClientEntityDelegate: https://gist.github.com/mhajas/bf6569316c5c53b73493508bbb674ca1

### CODE REVIEW

File: model/build-processor/src/main/java/org/keycloak/models/map/annotations/IgnoreForEntityImplementationGenerator.java [hunk 1]
```diff
@@ -25,7 +25,7 @@
  *
  * @author hmlnarik
  */
-@Retention(RetentionPolicy.SOURCE)
+@Retention(RetentionPolicy.CLASS)
```
Comment (mhajas) [inline model/build-processor/src/main/java/org/keycloak/models/map/annotations/IgnoreForEntityImplementationGenerator.java:28]: I was not able to make this work without changing this `RetentionPolicy`. With `RetentionPolicy.SOURCE` [this function](https://github.com/keycloak/keycloak/pull/8948/files#diff-f11d90d4ebaf77241f62c448873d53ce246c65fa54b7fe38cea6132efd7d02d2R107) doesn't work. It is also failing in GHA.

File: model/map-hot-rod/src/main/java/org/keycloak/models/map/common/AbstractHotRodEntity.java [hunk 2]
```diff
@@ -17,6 +17,5 @@
 
 package org.keycloak.models.map.common;
 
-public interface Versioned {
-    int getEntityVersion();
+public interface AbstractHotRodEntity {
```
Comment (mhajas) [inline model/map-hot-rod/src/main/java/org/keycloak/models/map/common/AbstractHotRodEntity.java:20]: I needed to mark all HotRodEntities with this interface in order to make this:  https://github.com/keycloak/keycloak/pull/8948/files#diff-1110818d9bc3b825b3324f67c809b938aebaf72c55fdbbf236736e17984b3591R376-R388 work.

File: model/map/src/main/java/org/keycloak/models/map/client/MapProtocolMapperEntity.java [hunk 3]
```diff
@@ -27,10 +28,7 @@
  */
 @GenerateEntityImplementations
 @DeepCloner.Root
-public interface MapProtocolMapperEntity extends UpdatableEntity {
-
-    String getId();
-    void setId(String id);
+public interface MapProtocolMapperEntity extends UpdatableEntity, AbstractEntity {
```
Comment (mhajas) [inline model/map/src/main/java/org/keycloak/models/map/client/MapProtocolMapperEntity.java:31]: I made MapProtocolMapperEntity interface extend AbstractEntity. It has method get/setId anyway. Do you see any reason why we don't want this?

File: model/map-hot-rod/src/main/java/org/keycloak/models/map/common/AbstractHotRodEntity.java (hunk 2 above)
Comment (mhajas) [reply model/map-hot-rod/src/main/java/org/keycloak/models/map/common/AbstractHotRodEntity.java:20]: Here is a link directly to the file: https://github.com/keycloak/keycloak/blob/dfa6636aa94372e4c7ca889e122c2f581414b7ff/model/build-processor/src/main/java/org/keycloak/models/map/processor/GenerateHotRodEntityImplementationsProcessor.java#L376-L388

With the previous link, it should be enough to scroll to the `GenerateHotRodEntityImplementationsProcessor` file and click `Show diff`.

**Review by martin-kanis — CHANGES_REQUESTED** [review]
Thank you @mhajas, great work! I mainly focused on actual output of the generator and compared it to the original `HotRodClientEntity`. It looks good to me. Infinispan indices still work.
I have only one nitpick, see inline.

**Comment by mhajas** [comment]
Thank you for the review @martin-kanis! I added the maven property for the annotation-api version

**Comment by martin-kanis** [comment]
> Thank you for the review @martin-kanis! I added the maven property for the annotation-api version

Thanks!

I have one more question. When I introduce a new field for lowercase variant i.e `clientIdLowercase`, Will it be possible to exclude generation of a setter and a getter for this field as we don't need them?

**Comment by mhajas** [comment]
Methods are generated based on `MapClientEntity` interface that should not contain lowercase fields so it shouldn't be a problem. You just need to declare setters, that are setting lowercase version of field, in [HotRodClientEntity.AbstractHotRodClientEntityDelegate](https://github.com/keycloak/keycloak/blob/ea8e461c32a66d01bf8672cacca2e1139d8ea5f9/model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/client/HotRodClientEntity.java#L147). When you declare a method in this class the processor will not attempt generate a declaration.

**Comment by martin-kanis** [comment]
> Methods are generated based on `MapClientEntity` interface that should not contain lowercase fields so it shouldn't be a problem. You just need to declare setters, that are setting lowercase version of field, in [HotRodClientEntity.AbstractHotRodClientEntityDelegate](https://github.com/keycloak/keycloak/blob/ea8e461c32a66d01bf8672cacca2e1139d8ea5f9/model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/client/HotRodClientEntity.java#L147). When you declare a method in this class the processor will not attempt generate a declaration.

Yeah thank you, I realized this after I asked the question :)

**Review by martin-kanis — DISMISSED** [review]

**Review by hmlnarik — CHANGES_REQUESTED** [review]
Thanks for the changes, looks very good!
Please see few comments / questions inline.

File: model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/client/HotRodClientEntity.java [hunk 4]
```diff
@@ -143,572 +144,40 @@
...
...thenticationFlowBindingOverrides) && Objects.equals(publicClient, entity.publicClient) && Objects.equals(fullScopeAllowed, entity.fullScopeAllowed) && Objects.equals(frontchannelLogout, entity.frontchannelLogout) && Objects.equals(notBefore, entity.notBefore) && Objects.equals(scope, entity.scope) && Objects.equals(webOrigins, entity.webOrigins) && Objects.equals(protocolMappers, entity.protocolMappers) && Objects.equals(clientScopes, entity.clientScopes) && Objects.equals(scopeMappings, entity.scopeMappings) && Objects.equals(surrogateAuthRequired, entity.surrogateAuthRequired) && Objects.equals(managementUrl, entity.managementUrl) && Objects.equals(baseUrl, entity.baseUrl) && Objects.equals(bearerOnly, entity.bearerOnly) && Objects.equals(consentRequired, entity.consentRequired) && Objects.equals(rootUrl, entity.rootUrl) && Objects.equals(standardFlowEnabled, entity.standardFlowEnabled) && Objects.equals(implicitFlowEnabled, entity.implicitFlowEnabled) && Objects.equals(directAccessGrantsEnabled, entity.directAccessGrantsEnabled) && Objects.equals(serviceAccountsEnabled, entity.serviceAccountsEnabled) && Objects.equals(nodeReRegistrationTimeout, entity.nodeReRegistrationTimeout);
```
[the remark this answers is not in the document]
Comment (mhajas) [reply model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/client/HotRodClientEntity.java:176]: Yes, I used Intellij for that

File: model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/common/HotRodEntityDescriptor.java [hunk 5]
```diff
@@ -17,35 +17,34 @@
...
+public class HotRodEntityDescriptor<HotRodEntityType, HotRodDelegate extends HotRodEntityDelegate<HotRodEntityType>> {
     private final Class<?> modelTypeClass;
-    private final Class<EntityType> entityTypeClass;
-    private final List<Class<?>> hotRodClasses;
+    private final Class<HotRodEntityType> entityTypeClass;
     private final String cacheName;
+    private final Function<HotRodEntityType, HotRodDelegate> hotRodDelegateProvider;
 
-    public HotRodEntityDescriptor(Class<?> modelTypeClass, Class<EntityType> entityTypeClass, List<Class<?>> hotRodClasses, String cacheName) {
+    public HotRodEntityDescriptor(Class<?> modelTypeClass, Class<HotRodEntityType> entityTypeClass, String cacheName, Function<HotRodEntityType, HotRodDelegate> hotRodDelegateProvider) {
```
[the remark this answers is not in the document]
Comment (mhajas) [reply model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/common/HotRodEntityDescriptor.java:28]: That sounds good, we can reuse it then.

File: model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/HotRodMapStorageProvider.java [hunk 6]
```diff
@@ -46,8 +47,8 @@ public HotRodMapStorageProvider(HotRodMapStorageProviderFactory factory, HotRodC
     }
 
     @SuppressWarnings("unchecked")
-    public <V extends AbstractEntity & UpdatableEntity, M> HotRodMapStorage<String, V, M> getHotRodStorage(Class<M> modelType, MapStorageProviderFactory.Flag... flags) {
-        HotRodEntityDescriptor<V> entityDescriptor = (HotRodEntityDescriptor<V>) factory.getEntityDescriptor(modelType);
+    public <V extends AbstractEntity & UpdatableEntity & HotRodEntityDelegate<E>, E, M> HotRodMapStorage<String, V, E, M> getHotRodStorage(Class<M> modelType, MapStorageProviderFactory.Flag... flags) {
```
[the remark this answers is not in the document]
Comment (mhajas) [reply model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/HotRodMapStorageProvider.java:50]: `V` needs to extend `HotRodEntityDelegate<E>` to be able to use `getHotRodEntity` when [adding stuff to the cache](https://github.com/keycloak/keycloak/blob/ea8e461c32a66d01bf8672cacca2e1139d8ea5f9/model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/HotRodMapStorage.java#L95).

`E` does not extend `V`. `E` represents HotRod entities. Maybe we can add `E extends AbstractHotRodEntity` to have type checking for this.

File: model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/common/HotRodPair.java [hunk 7]
```diff
@@ -49,4 +51,17 @@ public void setFirst(T first) {
     public void setSecond(V second) {
         this.secondWrapped = new WrappedMessage(second);
     }
+
```
[the remark this answers is not in the document]
Comment (mhajas) [reply model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/common/HotRodPair.java:54]: Sure

File: model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/common/HotRodEntityDescriptor.java [hunk 8]
```diff
@@ -17,35 +17,34 @@
 
 package org.keycloak.models.map.storage.hotRod.common;
 
-import java.util.List;
-import java.util.stream.Stream;
+import java.util.function.Function;
 
-public class HotRodEntityDescriptor<EntityType> {
+public class HotRodEntityDescriptor<HotRodEntityType, HotRodDelegate extends HotRodEntityDelegate<HotRodEntityType>> {
```
[the remark this answers is not in the document]
Comment (mhajas) [reply model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/common/HotRodEntityDescriptor.java:22]: Sure

File: model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/common/HotRodEntityDescriptor.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (hmlnarik) [reply model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/common/HotRodEntityDescriptor.java:28]: See #9025

**Comment by mhajas** [comment]
Thanks for the review @hmlnarik, I updated the PR based on your comments. Here is the new version of generated [HotRodClientEntityDelegate](https://gist.github.com/mhajas/4b2dcbbe054c34b34390fd434e0deba6).

File: model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/common/HotRodEntityDescriptor.java [hunk 9]
```diff
@@ -17,35 +17,34 @@
...
+public class HotRodEntityDescriptor<E, D extends HotRodEntityDelegate<E>> {
     private final Class<?> modelTypeClass;
-    private final Class<EntityType> entityTypeClass;
-    private final List<Class<?>> hotRodClasses;
-    private final String cacheName;
+    private final Class<E> entityTypeClass;
+    private final Function<E, D> hotRodDelegateProvider;
 
-    public HotRodEntityDescriptor(Class<?> modelTypeClass, Class<EntityType> entityTypeClass, List<Class<?>> hotRodClasses, String cacheName) {
+    public HotRodEntityDescriptor(Class<?> modelTypeClass, Class<E> entityTypeClass, Function<E, D> hotRodDelegateProvider) {
```
Comment (hmlnarik) [inline model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/common/HotRodEntityDescriptor.java:29]: There should be either `modelTypeClass` or `entityTypeClass` (perhaps the latter). The other one can be obtained from [`ModelEntityUtil`](https://github.com/keycloak/keycloak/blob/95614e8b4022306ffac328b85130403cc157fb18/model/map/src/main/java/org/keycloak/models/map/storage/ModelEntityUtil.java#L116)

Comment (mhajas) [reply model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/common/HotRodEntityDescriptor.java:29]: I am afraid that is not possible. `ModelEntityUtil` doesn't contain reference to HotRod entities, like `HotRodClientEntity`.

See: 
https://github.com/keycloak/keycloak/blob/7baaddd99dff29b139789009c9951c2f06a702c6/model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/HotRodMapStorageProviderFactory.java#L55-L58

**Review by hmlnarik — APPROVED** [review]

**Pull request merged**